File: core/x/javaapi/api/src/main/java/viaduct/java/api/types/Connection.java
package viaduct.java.api.types;

/**
 * A paginated list of {@link Edge}s plus a {@code pageInfo} object describing page boundaries.
 *
 * <p>Declare the type in your {@code .graphqls} schema — codegen generates the implementation and a
 * builder extending {@link viaduct.java.api.internal.ConnectionBuilder} with {@code fromSlice},
 * {@code fromList}, and {@code fromEdges}.
 *
 * <p>Java equivalent of Kotlin's {@code viaduct.api.types.Connection}.
 *
 * @param <E> the edge type wrapping nodes of type {@code N} and carrying a per-item cursor
 * @param <N> the node type returned by each edge
 * @see Edge
 * @see viaduct.java.api.internal.ConnectionBuilder
 */
public interface Connection<E extends Edge<N>, N> extends GraphQLObject {}

File: core/tenant/api/src/main/kotlin/viaduct/api/types/Connection.kt
package viaduct.api.types

import viaduct.apiannotations.ExperimentalApi

/**
 * A paginated list of [Edge]s plus a `pageInfo` object describing page boundaries.
 *
 * Declare the type in your `.graphqls` schema — codegen generates the implementation and a
 * builder extending [viaduct.api.internal.ConnectionBuilder] with [viaduct.api.internal.ConnectionBuilder.fromSlice],
 * [viaduct.api.internal.ConnectionBuilder.fromList], and [viaduct.api.internal.ConnectionBuilder.fromEdges].
 *
 * @param E The edge type wrapping nodes of type [N] and carrying a per-item cursor.
 * @param N The node type returned by each edge.
 * @see Edge
 * @see viaduct.api.internal.ConnectionBuilder
 */
@ExperimentalApi
interface Connection<E : Edge<N>, N> : Object
